File: btree_sql_java/src/main/java/com/btreesql/BTree.java
package com.btreesql;

import java.util.ArrayList;
import java.util.List;

/**
 * B-Tree implementation for indexing
 */
public class BTree {
    private static final int ORDER = 4; // Minimum degree

    private BTreeNode root;

    public BTree() {
        this.root = new BTreeNode(true);
    }

    public void insert(long key, long rowId) {
        BTreeNode r = root;

        if (r.isFull()) {
            BTreeNode newRoot = new BTreeNode(false);
            newRoot.children.add(r);
            splitChild(newRoot, 0);
            root = newRoot;
            insertNonFull(newRoot, key, rowId);
        } else {
            insertNonFull(r, key, rowId);
        }
    }

    private void insertNonFull(BTreeNode node, long key, long rowId) {
        int i = node.keys.size() - 1;

        if (node.isLeaf) {
            // Insert key in sorted order
            node.keys.add(0L);
            node.rowIds.add(0L);

            while (i >= 0 && key < node.keys.get(i)) {
                node.keys.set(i + 1, node.keys.get(i));
                node.rowIds.set(i + 1, node.rowIds.get(i));
                i--;
            }

            node.keys.set(i + 1, key);
            node.rowIds.set(i + 1, rowId);
        } else {
            // Find child to insert into
            while (i >= 0 && key < node.keys.get(i)) {
                i--;
            }
            i++;

            if (node.children.get(i).isFull()) {
                splitChild(node, i);
                if (key > node.keys.get(i)) {
                    i++;
                }
            }

            insertNonFull(node.children.get(i), key, rowId);
        }
    }

    private void splitChild(BTreeNode parent, int index) {
        BTreeNode child = parent.children.get(index);
        BTreeNode newNode = new BTreeNode(child.isLeaf);

        int midPoint = ORDER - 1;

        // Copy second half of keys to new node
        for (int i = midPoint + 1; i < child.keys.size(); i++) {
            newNode.keys.add(child.keys.get(i));
            newNode.rowIds.add(child.rowIds.get(i));
        }

        // If not a leaf, copy children too
        if (!child.isLeaf) {
            for (int i = midPoint + 1; i < child.children.size(); i++) {
                newNode.children.add(child.children.get(i));
            }
        }

        // Move middle key up to parent
        long middleKey = child.keys.get(midPoint);
        long middleRowId = child.rowIds.get(midPoint);

        // Trim child node
        if (!child.isLeaf) {
            child.children.subList(midPoint + 1, child.children.size()).clear();
        }
        child.keys.subList(midPoint, child.keys.size()).clear();
        child.rowIds.subList(midPoint, child.rowIds.size()).clear();

        // Insert into parent
        parent.keys.add(index, middleKey);
        parent.rowIds.add(index, middleRowId);
        parent.children.add(index + 1, newNode);
    }

    public Long search(long key) {
        return search(root, key);
    }

    private Long search(BTreeNode node, long key) {
        int i = 0;
        while (i < node.keys.size() && key > node.keys.get(i)) {
            i++;
        }

        if (i < node.keys.size() && key == node.keys.get(i)) {
            return node.rowIds.get(i);
        }

        if (node.isLeaf) {
            return null; // Not found
        }

        return search(node.children.get(i), key);
    }

    public void print() {
        System.out.println("B-Tree structure:");
        print(root, 0);
    }

    private void print(BTreeNode node, int level) {
        System.out.print("Level " + level + ": ");
        for (long key : node.keys) {
            System.out.print(key + " ");
        }
        System.out.println();

        if (!node.isLeaf) {
            for (BTreeNode child : node.children) {
                print(child, level + 1);
            }
        }
    }

    /**
     * B-Tree Node class
     */
    private static class BTreeNode {
        List<Long> keys;
        List<Long> rowIds;
        List<BTreeNode> children;
        boolean isLeaf;

        BTreeNode(boolean isLeaf) {
            this.keys = new ArrayList<>();
            this.rowIds = new ArrayList<>();
            this.children = new ArrayList<>();
            this.isLeaf = isLeaf;
        }

        boolean isFull() {
            return keys.size() >= 2 * ORDER - 1;
        }
    }
}

File: btree_sql_rust/src/btree.rs
use std::fmt;

const BTREE_ORDER: usize = 4;

#[derive(Debug, Clone)]
pub struct BTreeNode {
    keys: Vec<i64>,
    row_ids: Vec<i64>,
    children: Vec<Box<BTreeNode>>,
    is_leaf: bool,
}

impl BTreeNode {
    fn new(is_leaf: bool) -> Self {
        BTreeNode {
            keys: Vec::with_capacity(2 * BTREE_ORDER - 1),
            row_ids: Vec::with_capacity(2 * BTREE_ORDER - 1),
            children: Vec::with_capacity(2 * BTREE_ORDER),
            is_leaf,
        }
    }

    fn is_full(&self) -> bool {
        self.keys.len() >= 2 * BTREE_ORDER - 1
    }

    fn search(&self, key: i64) -> Option<i64> {
        let mut i = 0;
        while i < self.keys.len() && key > self.keys[i] {
            i += 1;
        }

        if i < self.keys.len() && key == self.keys[i] {
            return Some(self.row_ids[i]);
        }

        if self.is_leaf {
            return None;
        }

        self.children[i].search(key)
    }

    fn insert_non_full(&mut self, key: i64, row_id: i64) {
        let mut i = self.keys.len();

        if self.is_leaf {
            // Insert key in sorted order
            self.keys.push(0);
            self.row_ids.push(0);

            while i > 0 && key < self.keys[i - 1] {
                self.keys[i] = self.keys[i - 1];
                self.row_ids[i] = self.row_ids[i - 1];
                i -= 1;
            }

            self.keys[i] = key;
            self.row_ids[i] = row_id;
        } else {
            // Find child to insert into
            while i > 0 && key < self.keys[i - 1] {
                i -= 1;
            }

            if self.children[i].is_full() {
                self.split_child(i);
                if key > self.keys[i] {
                    i += 1;
                }
            }

            self.children[i].insert_non_full(key, row_id);
        }
    }

    fn split_child(&mut self, index: usize) {
        let mut new_node = BTreeNode::new(self.children[index].is_leaf);
        let child = &mut self.children[index];

        // Copy second half of keys to new node
        new_node.keys = child.keys.split_off(BTREE_ORDER - 1);
        new_node.row_ids = child.row_ids.split_off(BTREE_ORDER - 1);

        // Move middle key up to parent
        let middle_key = new_node.keys.remove(0);
        let middle_row_id = new_node.row_ids.remove(0);

        // If not a leaf, split children too
        if !child.is_leaf {
            new_node.children = child.children.split_off(BTREE_ORDER);
        }

        // Insert middle key into parent
        self.keys.insert(index, middle_key);
        self.row_ids.insert(index, middle_row_id);
        self.children.insert(index + 1, Box::new(new_node));
    }

    fn print(&self, level: usize) {
        print!("Level {}: ", level);
        for key in &self.keys {
            print!("{} ", key);
        }
        println!();

        if !self.is_leaf {
            for child in &self.children {
                child.print(level + 1);
            }
        }
    }
}

pub struct BTree {
    root: Box<BTreeNode>,
}

impl BTree {
    pub fn new() -> Self {
        BTree {
            root: Box::new(BTreeNode::new(true)),
        }
    }

    pub fn insert(&mut self, key: i64, row_id: i64) {
        if self.root.is_full() {
            let mut new_root = BTreeNode::new(false);
            let old_root = std::mem::replace(&mut *self.root, BTreeNode::new(true));
            new_root.children.push(Box::new(old_root));
            new_root.split_child(0);
            new_root.insert_non_full(key, row_id);
            self.root = Box::new(new_root);
        } else {
            self.root.insert_non_full(key, row_id);
        }
    }

    pub fn search(&self, key: i64) -> Option<i64> {
        self.root.search(key)
    }

    pub fn print(&self) {
        println!("B-Tree structure:");
        self.root.print(0);
    }
}

impl fmt::Display for BTree {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "BTree with {} keys", self.root.keys.len())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_insert_and_search() {
        let mut tree = BTree::new();
        tree.insert(10, 1);
        tree.insert(20, 2);
        tree.insert(30, 3);

        assert_eq!(tree.search(10), Some(1));
        assert_eq!(tree.search(20), Some(2));
        assert_eq!(tree.search(30), Some(3));
        assert_eq!(tree.search(40), None);
    }

    #[test]
    fn test_split() {
        let mut tree = BTree::new();
        for i in 1..=10 {
            tree.insert(i, i);
        }

        for i in 1..=10 {
            assert_eq!(tree.search(i), Some(i));
        }
    }
}
